### DISCUSSION TITLE
textfield keyboard not working on screen lock/unlock flutter

### DISCUSSION THREAD

**Discussion by GSPrakash2662** [community/community#73114] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Bug

### Body

I am working on a flutter application and having issue with textformfield. The behavior is when user tap on textfield (empty or entered some data) and lock the device and unlock it. after that textformfield won't take any input from keyboard. User had to again tap on textform field to enter the data.

Any solutions would be great.

And I have tried this piece of code also, But no luck:

`@override Future didChangeAppLifecycleState(AppLifecycleState state) async { if (state == AppLifecycleState.resumed) { FocusScope.of(context).unfocus(); } super.didChangeDependencies(); }
`

**Comment by iambhabha** [comment]
The issue you're describing could be related to the focus of the TextFormField not being restored properly when the app resumes after being in the background. While your attempt to unfocus the TextFormField in the didChangeAppLifecycleState is a good idea, it seems like it's not working as expected.

Here's an alternative approach you can try:

`class YourWidgetState extends State<YourWidget> with WidgetsBindingObserver {
  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addObserver(this);
  }

  @override
  void dispose() {
    WidgetsBinding.instance.removeObserver(this);
    super.dispose();
  }

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    super.didChangeAppLifecycleState(state);
    if (state == AppLifecycleState.resumed) {
      // This will force a rebuild and refocus the TextFormField
      setState(() {});
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: GestureDetector(
        onTap: () {
          // This will unfocus the TextFormField when tapping outside
          FocusScope.of(context).unfocus();
        },
        child: YourFormWidget(), // Replace with your TextFormField
      ),
    );
  }
}
`

In this approach, we're using WidgetsBindingObserver to listen for app lifecycle changes and force a rebuild of the widget when the app is resumed. This can help in refocusing the TextFormField.

Additionally, wrapping your widget with GestureDetector and handling the onTap event to unfocus when tapping outside the TextFormField might improve the overall user experience.

Remember to replace YourFormWidget() with the actual widget that contains your TextFormField. If the issue persists, you might need to investigate further or consider checking for updates in the Flutter framework or related packages.

**Comment by shahbazakon** [comment]
It seems like you've already tried handling the AppLifecycleState in your Flutter application, but the issue persists. While your approach is on the right track, it's possible that the problem might be related to the focus not being properly restored after the app is resumed.

Here's an alternative approach you can try:

```
class YourWidgetState extends State<YourWidget> with WidgetsBindingObserver {
  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance?.addObserver(this);
  }

  @override
  void dispose() {
    WidgetsBinding.instance?.removeObserver(this);
    super.dispose();
  }

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    super.didChangeAppLifecycleState(state);

    if (state == AppLifecycleState.resumed) {
      // Delay the focus restoration slightly to ensure it happens after the keyboard is fully initialized.
      Future.delayed(Duration(milliseconds: 300), () {
        FocusScope.of(context).requestFocus(FocusNode());
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    // Your widget implementation
  }
}

```

**Comment by samiulhaquereal** [comment]
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> with WidgetsBindingObserver {
  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addObserver(this);
  }

  @override
  void dispose() {
    WidgetsBinding.instance.removeObserver(this);
    super.dispose();
  }

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    if (state == AppLifecycleState.resumed) {
      SystemChannels.textInput.invokeMethod('TextInput.show');
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('TextFormField Issue'),
      ),
      body: Center(
        child: Padding(
          padding: EdgeInsets.all(20.0),
          child: TextFormField(
            decoration: InputDecoration(
              labelText: 'Enter text',
            ),
          ),
        ),
      ),
    );
  }
}

**Comment by Musaddiq625** [comment]
actually it's because of the focus node
when we lock/unlock the device the focus node gets lost, thats why keyboard not working

to fix this as a user experience, you can close the keyboard when the AppLifeCycle state is hidden 🤷🏻‍♂️
